File: testdata/safe/safe_app.go
package main

import (
	"crypto/sha256"
	"database/sql"
	"net/http"
)

func safeFetch() { _, _ = http.Get("http://internal/reports") } // constant URL

func strongHash(b []byte) [32]byte { return sha256.Sum256(b) } // sha256

func safeQuery(db *sql.DB, name string) {
	db.Exec("SELECT * FROM users WHERE name = $1", name) // parameterized
}

File: testdata/safe/safe_app.py
import hashlib
import requests
from flask import redirect

ALLOWED = {"reports": "http://internal/reports"}


def fetch(key):
    # Allow-listed URL, no user concatenation — not SSRF.
    return requests.get(ALLOWED[key])


def strong_hash(data):
    return hashlib.sha256(data).hexdigest()  # not weak crypto


def go_home():
    return redirect("/home")  # constant target — not an open redirect


def safe_query(db, name):
    db.execute("SELECT * FROM users WHERE name = %s", (name,))  # parameterized
